### ISSUE TITLE
Add support for custom certificate

### ISSUE THREAD

**Issue by kjlaw89** [electron/electron#1956]
This is related to the pull request #1859, which introduced the --client-certificate flag. When trying to use a PEM file in this manner it does not allow access to a site requesting the client certificate.

**Comment by kjlaw89** [comment]
Just incase it has any use, here's the SSL error that Electron throws:

    [6380:0612/085913:ERROR:ssl_client_socket_openssl.cc(1014)] handshake failed; returned 0, SSL error code 1, net_error -107
    [6380:0612/085913:ERROR:ssl_client_socket_openssl.cc(1014)] handshake failed; returned 0, SSL error code 1, net_error -157

**Comment by deepak1556** [comment]
Got around to finding the real issue with the certificate and test server sample provided by @kjlaw89 . Did a `log-net-log` trace on both chrome and Electron , found the following in socket connection stage after the server certificates were received.

```
t=1788 [st=1642]        SSL_CLIENT_CERT_REQUESTED
t=1788 [st=1642]        SSL_READ_ERROR
                        --> net_error = -110 (ERR_SSL_CLIENT_AUTH_CERT_NEEDED)
                        --> ssl_error = 2
```

there was no client certificate read, and `AtomBrowserClient::SelectClientCertificate` was never hit, turned out chromium does a initial search on the platfrom store for [non-emptyness](https://code.google.com/p/chromium/codesearch#chromium/src/content/browser/ssl/ssl_client_auth_handler.cc&sq=package:chromium&rcl=1434972555&l=163) before requesting the delegate on the UI thread to let user select certificate. I patched this initial check and `SelectClientCertificate` got triggered with data from PEM file passed back.  Now there is a second stage lookup on the platform store for private key in `openssl_platfrom_key_*.cc` , so the purpose of command line option is now obsolete as there is always a lookup on the platform store unless we patch it on chromium, which i dont suggest. Any suggestions to solve this problem @zcbenz

**Comment by zcbenz** [comment]
It seems that if we don't provide a cert store SelectClientCertificate will still be called?

**Comment by deepak1556** [comment]
Yup but then we cant do the fallback. Also if `SelectClientCertificate` is called there will still be a search on the key store for private key but that is being resolved in this CL https://code.google.com/p/chromium/issues/detail?id=394131 . So i guess command line option would be no go until then ?

**Comment by zcbenz** [comment]
> Also if SelectClientCertificate is called there will still be a search on the key store for private key but that is being resolved in this CL https://code.google.com/p/chromium/issues/detail?id=394131 . So i guess command line option would be no go until then ?

I think we can just wait for Chromium to support it then, we don't need to do better than Chromium.

**Comment by hokein** [comment]
I can verify it on other platforms.

**Comment by kjlaw89** [comment]
@deepak1556 Did Chrome 47 include the code necessary to get this flag to work? It'd be great if it was possible to use this now!

**Comment by kjlaw89** [comment]
@deepak1556 @zcbenz Okay, now that we're on Chrome 49, any hope yet? :)

**Comment by deepak1556** [comment]
looked into this further, currently there is no way to hook into a resource lifecycle where we can provide both certificate and private key. The only hook is when chromiums' resource loader asks for a certificate and it will look into the platform certificate database for the key. As far as i can see the current implementation of `client-certificate` flag doesnt serve its purpose and can be removed. Instead we can provide api to import certificate and ca authorities. Users can listen to `select-client-certificate` and proceed. @zcbenz thoughts ?

**Comment by zcbenz** [comment]
@deepak1556 Sounds good to me.

**Comment by kjlaw89** [comment]
@deepak1556 Thanks for looking into this again! What exactly are you proposing doing? I'd be okay with removing the client certificate flag if I had a way to dynamically load a client certificate. It was almost possible to do it with the select-client-certificate event, but it required that a client certificate be installed on the computer first before you could pass in a custom client certificate. If it was possible to get around that (no client certificate installed on the computer) and use my custom client certificate that'd be perfect.

**Comment by deepak1556** [comment]
> no client certificate installed on the computer)

Nope not possible to get around that currently.

My proposal was to provide api which would trigger the same action as you click the `import` button under certificate management in chrome. That way you can prompt users to install certificate. On linux you can get away without any prompts since only thing required there is password for the pkcs12 file, but not on OSX or windows. Will try to get it working in next few days.

**Comment by aarnaud** [comment]
Hi,

I have used  https://github.com/electron/electron/blob/master/docs/api/app.md#appimportcertificateoptions-callback-linux for importing client certificate. It works with signed certificate but not with self-signed certificate. 

How to import a self-signed certificate ?
How to list/manage imported certificates ?

Thanks,

**Comment by deepak1556** [comment]
@aarnaud the api should work for self-signed certificates, can you open a new issue with reproducible example ?

> How to list/manage imported certificates ?

You can use `select-client-certificate` event to choose the certificate, but deletion from cert store is not possible with the current api.

**Comment by aarnaud** [comment]
@deepak1556 I tried on a other computer with self-signed certificate without password it works.

**Comment by PatriqDesigns** [comment]
@deepak1556 Does app.importCertificate work on Windows and OS X? If so, electron keeps telling me there is no such function. If not is there any alternative? 

Thank you!

**Comment by jeandenis-k** [comment]
Hi @deepak1556, I'm ready to work on Mac/Windows implementation and testing. What would be your directions? Thank you for your help!

**Comment by deepak1556** [comment]
@jeandenis-k you can start by porting [settings_util::ShowManageSSLCertificates](https://cs.chromium.org/chromium/src/chrome/browser/ui/webui/settings_utils.h?l=29) for OSX and windows.

**Comment by lecogiteur** [comment]
Hello,

we need also these features. Do you have an idea for their availability with electron? Only on Windows, OSX, not on Linux? Does Self-signed certificate work on Linux?

Thanks

**Comment by jeandenis-k** [comment]
Hi! Import of SSL client certificates (in PKCS12 format) already works on linux. I'm working on Mac support, but it's slow going: on Linux, Chrome uses a custom certificate store (using the NSS lib), whereas on Mac/Win Chrome uses the system certificate store, which is complex and platform-specific.

As for self-signed certificates, I'm not really sure but looking at the implementation I don't see why that wouldn't work on Linux.

**Comment by lecogiteur** [comment]
@jeandenis-k Sorry, it's not worked on Linux for me. I try this [https://github.com/hokein/electron-sample-apps/tree/master/client-certificate](example).
I add in main.js
`app.importCertificate(....)`
 I have this error 
`[23447:0930/091053:ERROR:cert_verify_proc_nss.cc(942)] CERT_PKIXVerifyCert for 127.0.0.1 failed err=-8172
net::ERR_CERT_AUTHORITY_INVALID`

**Comment by deepak1556** [comment]
@lecogiteur the error denotes certificate was issued by an untrusted authority, which is true in your case. You should handle with `certificate-error` event of app or package the issuer with the p12 file, electron will import and set the appropriate trust bits.

**Comment by lecogiteur** [comment]
Thanks very much

@deepak1556 

> You should handle with certificate-error event 

I didn't say it, but I already try this. You are right, it fixes the problem of untrusted authority.

I finally found my error : 
In the main.js of [example](https://github.com/hokein/electron-sample-apps/tree/master/client-certificate), I must put the block code "app.importCertificate(....)" in the function associated to the event app `ready` before load URL. If I put it before the declaration of the `ready` event (out of the function), it doesn't work. The certificate is never imported. **Is it normal?**

I note with this block code:

`app.on('ready', function() {
  mainWindow = new BrowserWindow({
    'width': 800,
    'height': 600,
  });`
 ` var options = new Object();`
  `options.certificate="/XXXXXXX/client-certificate/ssl/client.p12";`
 ` options.password="a";`
  `app.importCertificate(options, () => (console.log("Imported")));`
  `console.log("load URL");`
  `mainWindow.loadURL('https://localhost:5000');`
  `console.log("end to load URL");`
`});`

The result in console is
`load URL`
`end to load URL`
`Imported`
`[8478:1001/094407:ERROR:cert_verify_proc_nss.cc(942)] CERT_PKIXVerifyCert for localhost failed err=-8172`
The certificate seems to be imported after to load URL.

**Comment by deepak1556** [comment]
Yes importing the certificate is async, you have wait for import result and then load your initial url. The other problem of api being available only after ready event is intentional, I see the documentation missed it, should fix it.

**Comment by lecogiteur** [comment]
Thanks. It's clear. 

Yes the documentation is quite light.

Sorry a last question for Windows and Electron 1.4.2: the list of certificates on `select-client-certificate` event contains the list of certificate already stored in Windows's certificate store (by another way using CryptoAPI for example)? Or It's not yet implemented for Windows?

**Comment by mashaalmemon** [comment]
Hi There,

I am using NightmareJS which relys on electron for web browsing. I have a situation where my proxy (specified via the --proxy-server switch) needs to use a self signed client certificate for https communication with the proxy. Without the client certificate it appears that "ERR_UNEXPECTED_PROXY_AUTH" is coming back from electron (my code etc at segmentio/nightmare#829).

It appears that the --client-certifcate switch was removed as it was not functional. Apparently the client certificate must be installed on the platform. A few questions about electron behaviour in contrast to curl behaviour:
- In curl I can do either of the following to get a curl request through the proxy. I can provide the client certificate to use:

```
// works with certificate
curl -x <proxy host>:<proxy port> -U <proxy user>:<proxy pass> --cacert /path/to/cert.pem https://www.google.com
```

alternately I can instruct curl to allow insecure ssl connections and transfers:

```
// works without certificate
curl -x <proxy host>:<proxy port> -U <proxy user>:<proxy pass> --insecure https://www.google.com
```

When using electron does the '--ignore-certificate-errors' switch have the same effect for client certificates? If not it may be prudent to add such a switch/feature to electron if possible. 

- In my environment we develop on Mac but deploy in production on linux. If my only work around here is to install the certificate in the platform storm then on Mac I assume I assume that in Mac the certificate is to be installed via the Keychain Access app. However should it be installed in the 'login' keychain or in a different keychain?

Thanks in advance for any feedback.

**Comment by Bluscream** [comment]
Can you add a option to add a cert as non-dev? so the user has a folder where he can drop certs?

**Comment by Inateno** [comment]
Hi, I have a question about this issue.
If we use a signed-certificate stored with the application to allow communication with our https servers.
What is the risk if the user takes the certificate and try to use it in others circumstances?

Just to be sure there is no "big risk" about giving a cert file to users...

**Comment by mdan312** [comment]
when I launch `bash setup.sh` from the ssl folder, I obtain: 
`e is 65537 (0x10001)
Subject does not start with '/'.
problems making Certificate Request`

I am using Windows 10 from GIT Bash command line.

I get the three .key files (rootCA.key, server.key and client.key) but no .crt, no .csr, no client.p12

**Comment by codebytere** [comment]
Closing as the original issue has been addressed.

**Comment by willstott101** [comment]
Should this still be open considering the limited platform support? Or can we re-open #7305 for instance?

**Comment by vraravam** [comment]
Could we please reopen either this one or #7305? Both of them being closed - while the support for non-linux OSes is missing, is dis-heartening.

**Comment by Tarrowren** [comment]
I use `session.defaultSession.webRequest.onBeforeRequest` to redirect to custom protocol and then use the `https.request` to validate server certificate and send client certificate. 
This allows you to manage your own certificates.

**Comment by oganix** [comment]
@Tarrowren do you have an example of that you can share?

**Comment by Tarrowren** [comment]
@oganix 
example: https://github.com/Tarrowren/electron-self-signed-certificate-example

**Comment by Bluscream** [comment]
Stop staling this, just fix it

**Comment by rahul-dutt-sharma** [comment]
This is a very valid use case for apps behind corporate networks. They use cloud proxies like zscaler, netlog to ssl inspect the traffic. It's becoming untenable to request them to exclude urls needed by our app from ssl inspection

**Comment by oleg-milantiev** [comment]
Have the same problem. Please help :)

We plan to using squid proxy with ssl-bump. Need to import root certificate into Electron App (Windows) for using this application in closed corporate network.

**Comment by James-Livesey** [comment]
Would like to see this implemented as I'm currently working on an application that will need a way for users to configure their certificates on Linux. Would be great to be able to manage certificates through CRUD methods to fully resemble Chrome's certificate store interface (in Settings > Privacy and security > Manage certificates). [`app.importCertificate`](https://www.electronjs.org/docs/latest/api/app#appimportcertificateoptions-callback-linux) is a start, but not being able to list certificates and remove them is rather limiting.

That being said, I've just had a brief look at [`ses.setCertificateVerifyProc`](https://www.electronjs.org/docs/latest/api/session#sessetcertificateverifyprocproc), which looks promising.

**Comment by Tarrowren** [comment]
I have refactored the code and it now looks simpler.
https://github.com/Tarrowren/electron-self-signed-certificate-example
Problem: I used 'undici' instead of 'net.fetch' because it doesn't support 'agent', which would cause a typescript type error, but I ignored it and now it works fine, but may also raise unexpected errors...
